spring-integration-core/src/main/java/org/springframework/integration/support/MessageBuilder.java

```diff
@@ -81,9 +81,15 @@ public T getPayload() {
 	 * @param <T> The type of the payload.
 	 * @return A MessageBuilder.
 	 */
+	@SuppressWarnings("unchecked")
 	public static <T> MessageBuilder<T> fromMessage(Message<T> message) {
 		Assert.notNull(message, "message must not be null");
-		return new MessageBuilder<T>(message.getPayload(), message);
+		if (message instanceof CloneableMessage) {
+			return ((CloneableMessage<T>) message).cloneMessage();
+		}
+		else {
+			return new MessageBuilder<T>(message.getPayload(), message);
+		}
```

Comment:
I thought about that, but there are many places when `MessageBuilder` is used directly.
And that would be disappointing from end-user perspective to see different behavior from factory and from builder.